src/operator/tensor/elemwise_binary_op.h

```diff
@@ -13,350 +13,729 @@
+          Kernel<BinaryOpMissingLValue<OP, Req>, xpu>::Launch(
+            s, rvalue.shape_.Size(), out[iter_out].dptr_, rvalue.dptr_);
+        });
+      }
+      iter_out++;
+    }
+    // Evaluate the remaining rows beyond the l and r value row intersetion
+    while (iter_l < num_rows_l) {
+      indices_out[iter_out] = indices_l[iter_l];
+      Tensor<xpu, 1, DType> lvalue = data_l[iter_l++];
+      MXNET_ASSIGN_REQ_SWITCH(req[0], Req, {
+        Kernel<BinaryOpMissingRValue<OP, Req>, xpu>::Launch(
+          s, lvalue.shape_.Size(), out[iter_out].dptr_, lvalue.dptr_);
+      });
+    }
+    while (iter_r < num_rows_r) {
+      indices_out[iter_out] = indices_r[iter_r];
+      Tensor<xpu, 1, DType> rvalue = data_r[iter_r++];
+      MXNET_ASSIGN_REQ_SWITCH(req[0], Req, {
+        Kernel<BinaryOpMissingLValue<OP, Req>, xpu>::Launch(
+          s, rvalue.shape_.Size(), out[iter_out].dptr_, rvalue.dptr_);
+      });
+    }
+    nnvm::TShape new_shape = output->aux_shape(rowsparse::kIdx);
+    new_shape[0] -= num_common_rows;  // Reduce the first-dimension size by the number of common rows
+    output->set_aux_shape(rowsparse::kIdx, new_shape);
+*/
+#endif
+    if(tempSparse.get()) {
+      // Required output is actually something other than RSP,
+      // so cast out final RSP to the true output type
+      CastStorageComputeImpl(s, *tempSparse, outputs[0]);
+    }
   }
-}
-
-// TODO(haibin) This is a single-thread inefficient implementation
-// Binary Compute between two row-sparse ndarray
-// This implementation only works on CPU
-template<typename xpu, typename OP>
-void BinaryComputeRspRsp(const nnvm::NodeAttrs& attrs,
-                         const OpContext& ctx,
-                         const std::vector<NDArray>& inputs,
-                         const std::vector<OpReqType>& req,
-                         const std::vector<NDArray>& outputs) {
-  auto &lhs = inputs[0];
-  auto &rhs = inputs[1];
-  auto &output = outputs[0];
-
-  bool init_l = lhs.storage_initialized();
-  bool init_r = rhs.storage_initialized();
-  // both inputs are zeros
-  if (!init_l && !init_r) return;
-  // one of the input is zeros
-  if (!init_l || !init_r) {
-    NDArray out(output);
-    CopyFromToRspImpl<xpu, xpu>(!init_l ? rhs : lhs, &out, ctx.run_ctx);
-    return;
+#endif
+
+  template<typename xpu, typename DType, typename OP>
+  static inline size_t FillDense(mshadow::Stream<xpu> *s,
+                                 const size_t idx_l,
+                                 const size_t idx_r,
+                                 const OpReqType req,
+                                 mshadow::Tensor<xpu, 2, DType>& out,
+                                 const size_t iter_out) {
+    using namespace mxnet_op;
+    using namespace mshadow::expr;
+    const int index_out_min = std::min(idx_l, idx_r);
+    const size_t size = out[iter_out].shape_.Size();
+    const DType zero_input_val = OP::Map(DType(0), DType(0));
+    #pragma omp parallel for
```

Comment:
This is for cpu only. But the function is xpu. Better to specialize the implementation and leave the gpu impl empty.